### PR TITLE
Allow Installed Apps to Display in the Game List

### PR DESCRIPTION
**Pull request by BreadFish64**
- Allows installed games to display in the game list
- Allows installed (executable) system applications to display in the game list
- Replaces the game icon with the icon from an installed update if available
image

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3402)

### CODE REVIEW

**Comment by valentinvanelslande** [comment]
the home menu does not appear

**Comment by BreadFish64** [comment]
@valentinvanelslande the home menu is technically an applet so it's not in the same directory as the other runnable system apps, I could add the six possible paths to detect it but I would want someone else's opinion on that

**Comment by chris062689** [comment]
I don't see the need to launch applets in the games list that's nonsensical.   @BreadFish64 I'd leave the home menu launching to another PR.

File: src/citra_qt/game_list.cpp [hunk 1]
```diff
@@ -401,8 +402,27 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, unsign
             u64 program_id = 0;
             loader->ReadProgramId(program_id);
 
+            std::vector<u8> update_smdh;
+            std::string update_path;
+            u64 update_id = 0;
+            std::unique_ptr<Loader::AppLoader> update_loader = nullptr;
```
Comment (MerryMage) [inline src/citra_qt/game_list.cpp]: Reduce the scope of these variables.

At least do:

```
            std::vector<u8> update_smdh;

            if (0x4000000000000 <= program_id && program_id <= 0x40000FFFFFFFF) {
                u64 update_id = program_id + 0xe00000000;
                std::string update_path =
                    Service::AM::GetTitleContentPath(Service::FS::MediaType::SDMC, update_id);
                if (FileUtil::Exists(update_path)) {
                    std::unique_ptr<Loader::AppLoader> update_loader = 
                        Loader::GetLoader(update_path);
                    if (update_loader) {
                        update_loader->ReadIcon(update_smdh);
                        update_loader->ReadProgramId(update_id);
                    }
                }
            }
```

Personally I would prefer the IIFE-style:

```
            std::vector<u8> update_smdh = [program_id]() -> std::vector<u8> {
                if (0x4000000000000 > program_id && program_id > 0x40000FFFFFFFF)
                    return {};

                u64 update_id = program_id + 0xe00000000;
                std::string update_path =
                    Service::AM::GetTitleContentPath(Service::FS::MediaType::SDMC, update_id);

                if (!FileUtil::Exists(update_path))
                    return {};

                std::unique_ptr<Loader::AppLoader> update_loader = Loader::GetLoader(update_path);

                if (!update_loader)
                    return {};

                std::vector<u8> update_smdh;
                update_loader->ReadIcon(update_smdh);
                update_loader->ReadProgramId(update_id);
                return update_smdh;
            }();
```

File: src/citra_qt/game_list.cpp [hunk 2]
```diff
@@ -421,7 +445,22 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, unsign
...
+                               "00000000000000000000000000000000/title/00040010"));
     AddFstEntriesToGameList(dir_path.toStdString(), deep_scan ? 256 : 0);
+    AddFstEntriesToGameList(
+        std::string(FileUtil::GetUserPath(D_SDMC_IDX).c_str()) +
+            "Nintendo "
+            "3DS/00000000000000000000000000000000/00000000000000000000000000000000/title/00040000",
+        true ? 256 : 0);
+    AddFstEntriesToGameList(std::string(FileUtil::GetUserPath(D_NAND_IDX).c_str()) +
+                                "00000000000000000000000000000000/title/00040010",
+                            true ? 256 : 0);
```
Comment (wwylele) [inline src/citra_qt/game_list.cpp]: what's up with the `true ? 256 : 0`? isn't it just `256`? Also, the directory hierarchy under sdmc or nand is well-defined and the recursion doesn't need to be greater than 5 or something.

File: src/citra_qt/game_list.cpp [hunk 3]
```diff
@@ -401,8 +402,31 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, unsign
...
+                std::vector<u8> update_smdh;
+                update_loader->ReadIcon(update_smdh);
+                update_loader->ReadProgramId(update_id);
+                return update_smdh;
+            }();
+
             emit EntryReady({
-                new GameListItemPath(QString::fromStdString(physical_name), smdh, program_id),
+                new GameListItemPath(QString::fromStdString(physical_name), smdh, program_id,
+                                     update_smdh),
```
Comment (wwylele) [inline src/citra_qt/game_list.cpp]: Is it necessary to let `GameListItemPath` object handle the two smdh? why not just replacing the `smdh` with `update_smdh` here (when update is found)?

Comment (BreadFish64) [reply src/citra_qt/game_list.cpp]: The SDMH is also where the program name is taken from... I might want to check what happens if I take that from update_sdmh as well

File: src/citra_qt/game_list.cpp (hunk 2 above)
Comment (BreadFish64) [reply src/citra_qt/game_list.cpp]: Tbh, I just put that there because the original had it, I could try actually figuring out what it does

Comment (FearlessTobi) [reply src/citra_qt/game_list.cpp]: IIRC it originally was used for recursive GameList search. When you enable it in the Citra menu, it searches Games in up to 256 subdirectories. I think you just copied it and replaced the boolean which enables recursion with true.

Comment (BreadFish64) [reply src/citra_qt/game_list.cpp]: Wwylele explain it in discord, I get what the problem is now :)

**Comment by LogPod** [comment]
Would be nice if we could have an option like say open the location where the game/app is installed.

File: src/citra_qt/game_list_p.h [hunk 4]
```diff
@@ -77,19 +77,19 @@ class GameListItemPath : public GameListItem {
...
         setData(qulonglong(program_id), ProgramIdRole);
 
+        Loader::SMDH smdh;
         if (!Loader::IsValidSMDH(smdh_data)) {
             // SMDH is not valid, set a default icon
             setData(GetDefaultIcon(true), Qt::DecorationRole);
             return;
         }
 
-        Loader::SMDH smdh;
```
Comment (wwylele) [inline src/citra_qt/game_list_p.h]: unnecessary change

**Comment by MerryMage** [comment]
Gonna merge this tmr if there are no complains.

**Comment by jroweboy** [comment]
:shipit:  Would like a follow up PR for what Logpod suggested as well if you get the chance.

**Comment by BreadFish64** [comment]
@jroweboy already started on it ;)
